share/functions/history.fish

```diff
@@ -2,93 +2,81 @@
 # Wrap the builtin history command to provide additional functionality.
 #
 function history --shadow-builtin --description "display or manipulate interactive command history"
-    # no args or just -t? use pager if interactive.
-    set -l cmd search
-    set -l prefix_args ""
-    set -l contains_args ""
-    set -l search_mode none
-    set -l time_args
+    set -l cmd
+    set -l search_mode --contains
+    set -l with_time
 
-    for i in (seq (count $argv))
-        if set -q argv[$i]
-            switch $argv[$i]
-                case -d --delete
-                    set cmd delete
-                case -v --save
-                    set cmd save
-                case -l --clear
-                    set cmd clear
-                case -s --search
-                    set cmd search
-                case -m --merge
-                    set cmd merge
-                case -h --help
-                    set cmd help
-                case -t --with-time
-                    set time_args "-t"
-                case -p --prefix
-                    set search_mode prefix
-                    set prefix_args $argv[(math $i + 1)]
-                case -c --contains
-                    set search_mode contains
-                    set contains_args $argv[(math $i + 1)]
-                case --
-                    set -e argv[1..$i]
-                    break
-                case "-*" "--*"
-                    printf ( _ "%s: invalid option -- %s\n" ) history $argv[$i] >&2
-                    return 1
-            end
+    # The "set cmd $cmd xyz" lines are to make it easy to detect if the user specifies more than one
+    # subcommand.
+    while set -q argv[1]
+        switch $argv[1]
+            case -d --delete
+                set cmd $cmd delete
+            case -v --save
+                set cmd $cmd save
+            case -l --clear
+                set cmd $cmd clear
+            case -s --search
+                set cmd $cmd search
+            case -m --merge
+                set cmd $cmd merge
+            case -h --help
+                set cmd $cmd help
+            case -t --with-time
+                set with_time -t
+            case -p --prefix
+                set search_mode --prefix
+            case -c --contains
+                set search_mode --contains
+            case --
+                set -e argv[1]
+                break
+            case '*'
```

Comment:
Something else that people might reasonably expect: `--prefix=foo` and `--contains=foo` to work.